### PR TITLE
Fixes for phpstormhelper to do with re-running failed tests

### PR DESCRIPTION
**Pull request by billypoke**
I have been getting console messages caused by the previous code I contributed when using the "re-run failed tests" button. This PR resolves those errors. I would really appreciate another set of eyes on this as well, as some of the removed code didn't seem to serve a purpose and I wanted to make sure there wasn't a use-case I was missing.

Also re-write for the util test to have more documentation and better-explained test cases

### CODE REVIEW

File: test/Unit/Util/PhpstormHelperTest.php [hunk 1]
```diff
@@ -146,75 +192,77 @@ public function providePhpstormCases(): Generator
...
+         * needs to be able to locate it elsewhere
+         */
         $argv   = [];
-        $argv[] = $paratestBinary;
+        $argv[] = $phpStormHelperBinary;
         $argv[] = sprintf('%s/bin/phpunit', uniqid());
-
-        $argv[] = '--runner';
-        $argv[] = 'WrapperRunner';
-        $argv[] = '--no-coverage';
```
Comment (Slamdunk) [inline test/Unit/Util/PhpstormHelperTest.php:199]: Why these options have been removed from this test case?

**Review by Slamdunk — CHANGES_REQUESTED** [review]

File: test/Unit/Util/PhpstormHelperTest.php (hunk 1 above)
Comment (steve-todorov) [reply test/Unit/Util/PhpstormHelperTest.php:199]: Most likely it's related to PhpStorm itself. This is the output you get when you re-run failed tests and add debugging:

```
/application/vendor/brianium/paratest/src/Util/PhpstormHelper.php:40:
array(10) {
  [0] =>  string(34) "./vendor/bin/paratest_for_phpstorm"
  [1] =>  string(24) "/application/bin/phpunit"
  [2] =>  string(15) "--configuration"
  [3] =>  string(29) "/application/phpunit.xml.dist"
  [4] =>  string(8) "--filter"
  [5] =>  string(111) "/(App\\Tests\\Services\\MyService::testSomething)( .*)?$/"
  [6] =>  string(13) "--test-suffix"
  [7] =>  string(24) "MyService.php"
  [8] =>  string(40) "/application/tests/Services/MyService"
  [9] =>  string(10) "--teamcity"
}
PHP Fatal error:  Uncaught RuntimeException: Missing path to 'vendor/brianium/paratest/bin/paratest' in /application/vendor/brianium/paratest/src/Util/PhpstormHelper.php:67
Stack trace:
#0 /application/vendor/brianium/paratest/src/Util/PhpstormHelper.php(42): ParaTest\Util\PhpstormHelper::getArgvKeyFor()
#1 /application/vendor/brianium/paratest/bin/paratest_for_phpstorm(9): ParaTest\Util\PhpstormHelper::handleArgvFromPhpstorm()
#2 /application/vendor/bin/paratest_for_phpstorm(115): include('/application/ve...')
#3 {main}
  thrown in /application/vendor/brianium/paratest/src/Util/PhpstormHelper.php on line 67
ERROR: 255

Process finished with exit code 255
```
You can check my full comment here:  https://github.com/paratestphp/paratest/issues/695#issuecomment-1271889215

**Comment by steve-todorov** [comment]
Hey,

I just tried this PR, but it doesn't seem to be working as expected for me at least.
Re-running failed tests works, but when I try to start paratest for all tests it fails unexpectedly:

```
[docker-compose://[/projects/clients/project/docker-compose.yaml]:php/]:php ./vendor/bin/paratest_for_phpstorm /application/bin/phpunit --configuration /application/phpunit.xml.dist --teamcity
Testing started at 14:18 ч. ...
WARNING: The DB_ENV variable is not set. Defaulting to a blank string.
Creating project_php_run ... 
Creating project_php_run ... done#!/usr/bin/env php
ParaTest v6.6.4 upon PHPUnit 9.5.25 #StandWithUkraine

<?xml version="1.0" encoding="UTF-8"?>

<!-- https://phpunit.readthedocs.io/en/latest/configuration.html -->
<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:noNamespaceSchemaLocation="vendor/phpunit/phpunit/phpunit.xsd"
         backupGlobals="false"
         colors="true"
         bootstrap="tests/bootstrap.php"
         convertDeprecationsToExceptions="false"
         printerClass="\LimeDeck\Testing\Printer"
>
    <php>
        <ini name="display_errors" value="1" />
        <ini name="error_reporting" value="-1" />
        <server name="APP_ENV" value="test" force="true" />
        <server name="SHELL_VERBOSITY" value="-1" />
        <server name="SYMFONY_PHPUNIT_REMOVE" value="" />
        <server name="SYMFONY_PHPUNIT_VERSION" value="9.5" />
    </php>

    <testsuites>
        <testsuite name="Project Test Suite">
            <directory>tests</directory>
        </testsuite>
    </testsuites>

    <coverage processUncoveredFiles="true">
        <include>
            <directory suffix=".php">src</directory>
        </include>
        <exclude>
            <directory>src/DataFixtures</directory>
            <directory>src/Doctrine</directory>
        </exclude>
    </coverage>

    <listeners>
        <listener class="Symfony\Bridge\PhpUnit\SymfonyTestsListener" />
        <listener class="Symfony\Bridge\PhpUnit\CoverageListener"/>
    </listeners>

    <!-- Run `composer require symfony/panther` before enabling this extension -->
    <!--
    <extensions>
        <extension class="Symfony\Component\Panther\ServerExtension" />
    </extensions>
    -->
</phpunit>

Time: 00:00.032, Memory: 6.00 MB

OK (0 tests, 0 assertions)
Process finished with exit code 0
```

It shows `0 tests` and `0 assertions` and prints out the `phpunit.xml` file which is odd.

With the patch I suggested in https://github.com/paratestphp/paratest/issues/695#issuecomment-1271889215 - everything works for me - both starting paratest for all tests and then re-running failed tests. Unfortunately I don't have the time to write the tests for it right now. If you are interested in writing them - here is the full code that works for me in both cases:

<details>
    <summary>ParaTest\Util\PhpstormHelper.php</summary>

      <?php
      
      declare(strict_types=1);
      
      namespace ParaTest\Util;
      
      use RuntimeException;
      
      use function array_search;
      use function array_unshift;
      use function in_array;
      use function strlen;
      use function substr_compare;
      
      /** @internal */
      final class PhpstormHelper
      {
          /** @param  array<int, string> $argv */
          public static function handleArgvFromPhpstorm(array &$argv, string $paratestBinary): string
          {
              $phpunitKey = self::getArgvKeyFor($argv, '/phpunit');
      
              if (! in_array('--filter', $argv, true)) {
                  $coverageArgKey = self::getCoverageArgvKey($argv);
                  if ($coverageArgKey !== false) {
                      unset($argv[$coverageArgKey]);
                  }
      
                  unset($argv[$phpunitKey]);
      
                  return $paratestBinary;
              }
      
              $packagePath = dirname(dirname(__DIR__));
              $binPath = $packagePath.DIRECTORY_SEPARATOR.'bin'.DIRECTORY_SEPARATOR.'paratest_for_phpstorm';
              $binPath = "vendor".DIRECTORY_SEPARATOR.substr($binPath, strlen($packagePath)+1);
      
              if(realpath($binPath) == false) {
                  throw new RuntimeException("Cannot resolve absolute path to ./bin/paratest_for_phpstorm");
              }
      
              unset($argv[self::getArgvKeyFor($argv, $binPath)]);
              $phpunitBinary = $argv[$phpunitKey];
              foreach ($argv as $index => $value) {
                  if ($value === '--configuration' || $value === '--bootstrap') {
                      break;
                  }
      
                  unset($argv[$index]);
              }
      
              array_unshift($argv, $phpunitBinary);
      
              return $phpunitBinary;
          }
      
          /** @param  array<int, string> $argv */
          private static function getArgvKeyFor(array $argv, string $searchFor): int
          {
      
              foreach ($argv as $key => $arg) {
                  if (self::strEndsWith($arg, $searchFor)) {
                      return $key;
                  }
              }
      
              throw new RuntimeException("Missing path to '$searchFor'");
          }
      
          /**
           * Polyfill from PHP 8.0, drop when 7.4 support ends
           */
          public static function strEndsWith(string $haystack, string $needle): bool
          {
      
              if ($needle === '' || $needle === $haystack) {
                  return true;
              }
      
              if ($haystack === '') {
                  return false;
              }
      
              $needleLength = strlen($needle);
      
              return $needleLength <= strlen($haystack) && substr_compare($haystack, $needle, -$needleLength) === 0;
          }
      
          /**
           * @param  array<int, string> $argv
           *
           * @return int|false
           */
          private static function getCoverageArgvKey(array $argv)
          {
              $coverageOptions = [
                  '-dpcov.enabled=1',
                  '-dxdebug.mode=coverage',
              ];
      
              foreach ($coverageOptions as $coverageOption) {
                  $key = array_search($coverageOption, $argv, true);
                  if ($key !== false) {
                      return $key;
                  }
              }
      
              return false;
          }
      }
      

</details>

File: test/Unit/Util/PhpstormHelperTest.php (hunk 1 above)
Comment (billypoke) [reply test/Unit/Util/PhpstormHelperTest.php:199]: @Slamdunk this test is for phpunit in an unexpected location. My thought was that each test should only use the options needed for that test. Here the wrapper is not one of the options being tested, so I removed it.

**Comment by billypoke** [comment]
@steve-todorov can you check the latest commit on this branch to see if you are still getting the dump of phpunit.xml? I don't think all the backtracking though dirname is needed, we just need to remove the argv associated with the helper :thinking:

**Comment by steve-todorov** [comment]
Hey! Yep, this seems to be working for me in both cases (in Linux, not sure for Windows as I don't have one). :)

**Review by Slamdunk — APPROVED** [review]
I'm not fully aware of the consequences of this PR, but I trust you this change is for the better

**Pull request merged**